### PR TITLE
Make Node.js exists function validate its argument

### LINKED ISSUE #2094: [Bug] Exception when calling fs.exists with undefined
- [x] I'd be willing to implement a fix

**Describe the bug**

Calling `fs.exists(undefined)` throws an exception when the call of the function is wrapped by yarn. When calling it without yarn, node just returns `false`.

**To Reproduce**

<details>
  <summary>Reproduction</summary>

  ```js repro
const fs = require('fs')
await fs.exists(undefined)
  ```

Or with node repl:
```
$ yarn node
Welcome to Node.js v12.19.0.
Type ".help" for more information.
> const fs = require('fs')
undefined
> fs.existsSync(undefined)
false
> fs.exists(undefined, r => { console.log(r) })
undefined
> Uncaught TypeError: Cannot read property 'match' of undefined
    at Function.resolveVirtual (/home/bfr/projects/catherine-jauffret/.pnp.js:27889:21)
    at VirtualFS.mapToBase (/home/bfr/projects/catherine-jauffret/.pnp.js:27926:22)
    at VirtualFS.existsPromise (/home/bfr/projects/catherine-jauffret/.pnp.js:27651:43)
    at PosixFS.existsPromise (/home/bfr/projects/catherine-jauffret/.pnp.js:27651:24)
>
```
</details>

**Environment if relevant (please complete the following information):**

 - OS: All
 - Node version: 10.x, 12.x, 14.x, 15.x
 - Yarn version: 2.3.3

### PR DESCRIPTION
**Pull request by BertrandFritsch**
**What's the problem this PR addresses?**

Make the `fs.exists` function mimic the behavior of the node function regarding to its behavior when the function is called with anything else that a string, especially when it gets `undefined`. There are libraries out there that call `fs.exists` with an environment variable without first checking whether the variable is defined.

Resolves #2094 

**How did you fix it?**

Validate the argument at the beginning of the function.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by arcanis** [comment]
Hey! Thanks for looking into that 👍

I don't think we want the `FakeFS` interface (which is intended to be type-safe) to expose this weird behaviour. Instead, the fix would be to encode it within the patch layer (similar to what we do [here](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-fslib/sources/index.ts#L147) and [here](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-fslib/sources/index.ts#L203)).

File: .yarn/versions/23967d0d.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+releases:
```
Comment (arcanis) [inline .yarn/versions/23967d0d.yml:1]: You don't need to mark all packages for release - just mark both the fslib and the CLI as patch releases, and set everything else on "decline".

Comment (merceyz) [reply .yarn/versions/23967d0d.yml:1]: Well `@yarnpkg/plugin-pnp` and `@yarnpkg/pnp` should be released as well since the PnP hook needs to be rebuilt

Comment (BertrandFritsch) [reply .yarn/versions/23967d0d.yml:1]: @arcanis @merceyz, I made the requested changes in the code. Note the sync version of `fs.exists` had already a catch all exception attached, so I left that function as is.

**Comment by BertrandFritsch** [comment]
@arcanis What about the failure of the Testing chores check? Should I add the `plugin-interactive-tools` and `plugin-typescript` to the packages to release too?

**Comment by merceyz** [comment]
Mark them as declined, after that you need to update the PnP hook to make it take effect, run `yarn build:pnp:hook && yarn` and commit the result

**Comment by BertrandFritsch** [comment]
@merceyz I did what you asked me to do, but still some checks are failing. Anything else that I can do, or did I something wrong?

**Comment by merceyz** [comment]
> but still some checks are failing. Anything else that I can do, or did I something wrong?

You had a merge conflict / your branch was out of date with master.

File: packages/yarnpkg-fslib/tests/patchedFs.test.ts [hunk 2]
```diff
@@ -30,4 +30,13 @@ describe(`patchedFs`, () => {
       done();
     });
   });
+
+  it(`in case of the parameter of fs.exists is not a string, give false`, done => {
+    const patchedFs = extendFs(fs, new PosixFS(new NodeFS()));
+
+    patchedFs.exists(undefined as any, exists => {
```
Comment (merceyz) [inline packages/yarnpkg-fslib/tests/patchedFs.test.ts:37]: Shouldn't the sync version also have this behaviour?

Comment (BertrandFritsch) [reply packages/yarnpkg-fslib/tests/patchedFs.test.ts:37]: Actually, it already had it. There is a catch all exception handler that just returns false.

**Comment by arcanis** [comment]
I've updated your diff, will land once the tests are green! As it turns out, `exists` was correctly patched (since we were catching rejections to return `false`) - the problem was that the error was an exception, not a rejection, which occurred because various promise methods from `ProxiedFS` were lacking the `async` modifier.

**Pull request merged**